File: LegacySource/settlement/room/main/employment/RoomEmployment.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace Settlement.Room.Main.Employment
{
    public static class ROOMS
    {
        public static Employment Employment => new Employment();
    }

    public class Employment
    {
        public Employer Employer { get; } = new Employer();
    }

    public class Employer
    {
        public void UpdateAll() { }
    }

    public class HistoryInt
    {
        private readonly int[] _data;
        private readonly int _capacity;
        private int _index;

        public HistoryInt(int capacity, int days, bool circular)
        {
            _capacity = capacity;
            _data = new int[capacity];
        }

        public void Inc(int value)
        {
            _data[_index] += value;
            _index = (_index + 1) % _capacity;
        }

        public void Save(FilePutter file)
        {
            file.Write(_data);
        }

        public void Load(FileGetter file)
        {
            file.Read(_data);
        }

        public void Clear()
        {
            Array.Fill(_data, 0);
        }
    }

    public class FilePutter
    {
        private readonly BinaryWriter _writer;

        public FilePutter(BinaryWriter writer)
        {
            _writer = writer;
        }

        public void I(int value)
        {
            _writer.Write(value);
        }

        public void Is(int[] values)
        {
            _writer.Write(values.Length);
            _writer.Write(values);
        }
    }

    public class FileGetter
    {
        private readonly BinaryReader _reader;

        public FileGetter(BinaryReader reader)
        {
            _reader = reader;
        }

        public int I()
        {
            return _reader.ReadInt32();
        }

        public int[] Is()
        {
            int length = _reader.ReadInt32();
            return _reader.ReadInt32Array(length);
        }
    }

    public class CLAMP
    {
        public static int I(int value, int min, int max)
        {
            return Math.Min(max, Math.Max(min, value));
        }
    }

    public class ArrayListResize<T> : List<T>
    {
        public ArrayListResize(int initialCapacity, int maxCapacity) : base(initialCapacity)
        {
        }

        public int Add(T item)
        {
            base.Add(item);
            return Count - 1;
        }

        public void Clear()
        {
            base.Clear();
        }
    }

    public class INDEXED
    {
        public virtual int Index() => throw new NotImplementedException();
    }

    public class GRoupInt : INT_OE<WGROUP>
    {
        private int[] _racePrio;
        private int _total;
        private readonly int _min;
        private readonly int _max;

        public GRoupInt(int min, int max)
        {
            _min = min;
            _max = max;
            _racePrio = new int[WGROUP.All().Count];
        }

        public void Save(FilePutter file)
        {
            file.Is(_racePrio);
            file.I(_total);
        }

        public void Load(FileGetter file)
        {
            _racePrio = file.Is();
            _total = file.I();
        }

        public void Clear()
        {
            Array.Fill(_racePrio, 0);
            _total = 0;
        }

        public override int Max(WGROUP t) => _max;
        public override int Min(WGROUP t) => _min;
        public override int Get(WGROUP t) => t == null ? _total : _racePrio[t.Index()];
        public override void Set(WGROUP t, int i)
        {
            i = CLAMP.I(i, Min(t), Max(t));
            if (_racePrio[t.Index()] != i)
            {
                _total -= _racePrio[t.Index()];
                _racePrio[t.Index()] = i;
                _total += _racePrio[t.Index()];
            }
        }
    }

    public class Target
    {
        private int _target;
        private readonly int[] _perGroup;

        public Target(RoomEmployment p)
        {
            _perGroup = new int[WGROUP.All().Count];
        }

        public int Get() => _target;

        public int Group(WGROUP g) => _perGroup[g.Index()];

        public void Clear()
        {
            _target = 0;
            foreach (var e in WGROUP.All())
            {
                ROOMS.Employment.ChangeTarget(-_perGroup[e.Index()], e);
                _perGroup[e.Index()] = 0;
            }
        }

        public void Add(WGROUP g, int amount)
        {
            ROOMS.Employment.ChangeTarget(amount, g);
            _target += amount;
            _perGroup[g.Index()] += amount;
        }

        public void Save(FilePutter file)
        {
            file.I(_target);
            file.Is(_perGroup);
        }

        public void Load(FileGetter file)
        {
            _target = file.I();
            _perGroup = file.Is();
        }
    }

    public class Priority : INTE
    {
        private readonly RoomEmployment _p;
        private int _prio = 10;

        public Priority(RoomEmployment p)
        {
            _p = p;
        }

        public void Save(FilePutter file)
        {
            file.I(_prio);
        }

        public void Load(FileGetter file)
        {
            _prio = file.I();
        }

        public void Clear()
        {
            _prio = 10;
        }

        public override int Max() => 30;
        public override int Min() => 0;
        public override int Get() => _prio;
        public override void Set(int i)
        {
            i = CLAMP.I(i, Min(), Max());
            if (_prio != i)
            {
                _prio = i;
                ROOMS.Employment.Employer.UpdateAll();
            }
        }
    }

    public class RoomEmployment : RoomEmploymentSimple, INDEXED
    {
        private static readonly ArrayListResize<RoomEmployment> WORK = new ArrayListResize<RoomEmployment>(10, 512);

        static RoomEmployment()
        {
            new GameDisposable
            {
                Dispose = () => WORK.Clear()
            };
        }

        private readonly int _index = WORK.Add(this);
        private readonly HistoryInt _history = new HistoryInt(32, TIME.Days(), true);
        public readonly Target target = new Target(this);
        public readonly Priority priority = new Priority(this);
        public readonly GRoupInt priorities = new GRoupInt(0, 5);

        public RoomEmployment(RoomBlueprintIns<?> p, RoomInitData init) : base("WORK", p, init)
        {
        }

        public HISTORY_INT History() => _history;

        protected override void Employ(Humanoid h, int delta)
        {
            ROOMS.Employment.History.Inc(-Employed());
            base.Employ(h, delta);
            ROOMS.Employment.ChangeCurrent(delta, WGROUP.Get(h.Indu()));
            ROOMS.Employment.History.Inc(Employed());
            _history.Set(Employed());
        }

        protected override void Register(RoomEmploymentIns ins, int delta)
        {
            ROOMS.Employment.ChangeNeeded(Blueprint(), -WorkersNeeded);
            base.Register(ins, delta);
            ROOMS.Employment.ChangeNeeded(Blueprint(), WorkersNeeded);
        }

        public override void Save(FilePutter file)
        {
            base.Save(file);
            target.Save(file);
            priority.Save(file);
            priorities.Save(file);
            _history.Save(file);
        }

        public override void Load(FileGetter file)
        {
            base.Load(file);
            target.Load(file);
            priority.Load(file);
            priorities.Load(file);
            _history.Load(file);
        }

        public override void Clear()
        {
            base.Clear();
            target.Clear();
            priority.Clear();
            priorities.Clear();
            _history.Clear();
            SetPrioOnSkill();
        }

        public void SetPrioOnSkill()
        {
            foreach (var g in WGROUP.All())
            {
                SetPrioOnSkill(g);
            }
        }

        public void SetPrioOnFullfillment()
        {
            foreach (var g in WGROUP.All())
            {
                SetPrioOnFullfillment(g);
            }
        }

        public void SetPrioOnSkill(WGROUP g)
        {
            int p = CLAMP.I((int)Math.Round(RACES.Boosts().GetNorSkill(g.Race, this) * priorities.Max), 1, priorities.Max);
            priorities.Set(g, p);
        }

        public void SetPrioOnFullfillment(WGROUP g)
        {
            int p = CLAMP.I((int)Math.Round(g.Race.Pref().GetWork(this) * priorities.Max), 1, priorities.Max);
            priorities.Set(g, p);
        }

        public override int Index() => _index;

        public override double Efficiency()
        {
            return base.Efficiency();
        }
    }

    public abstract class RoomEmploymentSimple
    {
        protected RoomEmploymentSimple(string name, RoomBlueprintIns<?> p, RoomInitData init) { }

        protected abstract void Employ(Humanoid h, int delta);
        protected abstract void Register(RoomEmploymentIns ins, int delta);
        public abstract void Save(FilePutter file);
        public abstract void Load(FileGetter file) throws IOException;
        public abstract void Clear();
        public abstract double Efficiency();
    }

    public class GameDisposable
    {
        public Action Dispose { get; set; }
    }

    public class TIME
    {
        public static int Days() => 0; // Placeholder
    }

    public class RACES
    {
        public Boosts Boosts() => new Boosts();
    }

    public class Boosts
    {
        public double GetNorSkill(Race race, RoomEmploymentSimple roomEmployment) => 0; // Placeholder
    }

    public class Race
    {
        public Pref Pref() => new Pref();
    }

    public class Pref
    {
        public double GetWork(RoomEmploymentSimple roomEmployment) => 0; // Placeholder
    }

    public class WGROUP
    {
        public static IEnumerable<WGROUP> All() => new List<WGROUP>();
        public static WGROUP Get(int index) => new WGROUP();
        public Race Race { get; }
        public int Index() => 0; // Placeholder
    }

    public class Humanoid
    {
        public int Indu() => 0; // Placeholder
    }

    public class SETT
    {
        public static Rooms Rooms() => new Rooms();
    }

    public class Rooms
    {
        public Employment Employment => new Employment();
    }

    public class RoomBlueprintIns<T>
    {
    }

    public class RoomInitData
    {
    }

    public class RoomEmploymentIns
    {
    }

    public class INTE
    {
        public abstract int Max();
        public abstract int Min();
        public abstract int Get();
        public abstract void Set(int i);
    }

    public class INT_OE<T>
    {
        public abstract int Max(T t);
        public abstract int Min(T t);
        public abstract int Get(T t);
        public abstract void Set(T t, int i);
    }
}

File: settlement/room/main/employment/RoomEmployment.java
package settlement.room.main.employment;

import static settlement.main.SETT.ROOMS;

import java.io.IOException;
import java.util.Arrays;

import game.GameDisposable;
import game.time.TIME;
import init.race.RACES;
import init.type.WGROUP;
import settlement.entity.humanoid.Humanoid;
import settlement.main.SETT;
import settlement.room.main.RoomBlueprintIns;
import settlement.room.main.util.RoomInitData;
import snake2d.util.file.Alloc;
import snake2d.util.file.FileGetter;
import snake2d.util.file.FilePutter;
import snake2d.util.misc.CLAMP;
import snake2d.util.sets.ArrayListResize;
import snake2d.util.sets.INDEXED;
import util.data.INT.INTE;
import util.data.INT_O.INT_OE;
import util.statistics.HISTORY_INT;
import util.statistics.HistoryInt;

public final class RoomEmployment extends RoomEmploymentSimple implements INDEXED{

	static ArrayListResize<RoomEmployment> WORK = new ArrayListResize<>(10, 512);
	static {
		new GameDisposable() {
			@Override
			protected void dispose() {
				WORK.clear();
			}
		};
	}
	
	private final int index = WORK.add(this);
	private final HistoryInt history = new HistoryInt(32, TIME.days(), true);
	public final Target target = new Target(this);
	public final Priority priority = new Priority(this);
	public final GRoupInt priorities = new GRoupInt(0, 5) {
		
		@Override
		public void set(WGROUP t, int i) {
			int o = get(t);
			super.set(t, i);
			if (o != get(t))
				SETT.ROOMS().employment.employer.updateAll();
		};
	};
	
	public RoomEmployment(RoomBlueprintIns<?> p, RoomInitData init) {
		super("WORK", p, init);
		
	}
	
	
	public HISTORY_INT history() {
		return history;
	}
	
	@Override
	void employ(Humanoid h, int delta) {
		SETT.ROOMS().employment.history.inc(-employed());
		super.employ(h, delta);
		ROOMS().employment.changeCurrent(delta, WGROUP.get(h.indu()));
		SETT.ROOMS().employment.history.inc(employed());
		history.set(employed());
	}
	
	@Override
	void register(RoomEmploymentIns ins, int delta) {
		ROOMS().employment.changeNeeded(blueprint(), -workersNeeded);
		super.register(ins, delta);
		ROOMS().employment.changeNeeded(blueprint(), workersNeeded);
	}
	
//	@Override
//	void fixFuckup() {
//		ROOMS().employment.changeNeeded(blueprint(), -workersNeeded);
//		super.fixFuckup();
//	}
	

	@Override
	void save(FilePutter file) {
		super.save(file);
		target.save(file);
		priority.save(file);
		priorities.save(file);
		history.save(file);
	}

	@Override
	void load(FileGetter file) throws IOException {
		super.load(file);
		target.load(file);
		priority.load(file);
		priorities.load(file);
		history.load(file);
	}

	@Override
	void clear() {
		super.clear();
		target.clear();
		priority.clear();
		priorities.clear();
		history.clear();
		setPrioOnSkill();
		
	}
	
	public void setPrioOnSkill() {
		for (WGROUP g : WGROUP.all()) {
			setPrioOnSkill(g);
		}
	}
	
	public void setPrioOnFullfillment() {
		for (WGROUP g : WGROUP.all()) {
			setPrioOnFullfillment(g);
		}
	}
	
	public void setPrioOnSkill(WGROUP g) {
		int p = CLAMP.i((int)Math.round(RACES.boosts().getNorSkill(g.race, this)*priorities.max), 1, priorities.max);
		priorities.set(g, p);
	}
	
	public void setPrioOnFullfillment(WGROUP g) {
		int p = CLAMP.i((int)Math.round(g.race.pref().getWork(this)*priorities.max), 1, priorities.max);
		priorities.set(g, p);
	}

	@Override
	public int index() {
		return index;
	}
	
	@Override
	public double efficiency() {
		// TODO Auto-generated method stub
		return super.efficiency();
	}
	
	public static class Priority implements INTE{
		
		final RoomEmployment p;
		private int prio = 10;

		public Priority(RoomEmployment p) {
			this.p = p;
		}
		
		void save(FilePutter file) {
			file.i(prio);
		}

		void load(FileGetter file) throws IOException {
			prio = file.i();
		}

		void clear() {
			prio = 10;
		}
		
		@Override
		public int max() {
			return 30;
		};
		
		@Override
		public int min() {
			return 0;
		}

		@Override
		public int get() {
			return prio;
		}

		@Override
		public void set(int i) {
			i = CLAMP.i(i, min(), max());
			if (prio != i) {
				prio = i;
				SETT.ROOMS().employment.employer.updateAll();
			}
		};
	}
	
	public static class GRoupInt implements INT_OE<WGROUP>{
		
		private int total;

		private final int[] racePrio = Alloc.ii(WGROUP.all().size());
		private final int min;
		private final int max;
		
		public GRoupInt(int min, int max) {
			this.min = min;
			this.max = max;
		}
		
		void save(FilePutter file) {
			file.is(racePrio);
			file.i(total);
		}

		void load(FileGetter file) throws IOException {
			file.is(racePrio);
			total = file.i();
		}

		void clear() {
			Arrays.fill(racePrio, 0);
			total = 0;
		}
		
		@Override
		public int max(WGROUP t) {
			return max;
		};
		
		@Override
		public int min(WGROUP t) {
			return min;
		}

		@Override
		public int get(WGROUP t) {
			if (t == null)
				return total;
			return racePrio[t.index()];
		}

		@Override
		public void set(WGROUP t, int i) {
			i = CLAMP.i(i, min(t), max(t));
			if (racePrio[t.index()] != i) {
				
				total -= racePrio[t.index()];
				racePrio[t.index()] = i;
				total += racePrio[t.index()];
				
			}
		};
		
	}
	
	public static class Target {
		
		private int target;
		private final int[] perGroup = Alloc.ii(WGROUP.all().size());

		
		public Target(RoomEmployment p) {

		}
		
		public int get() {
			return target;
		}
		
		public int group(WGROUP g) {
			return perGroup[g.index()];
		}
		
		void clear() {
			target = 0;
			for (WGROUP e : WGROUP.all()) {
				SETT.ROOMS().employment.changeTarget(-perGroup[e.index()], e);
				perGroup[e.index()] = 0;
			}
			
		}
		
		void add(WGROUP g, int amount) {
			SETT.ROOMS().employment.changeTarget(amount, g);
			target+= amount;
			perGroup[g.index()] += amount;
		}
		

		
		void save(FilePutter file) {
			file.i(target);
			file.is(perGroup);
		}

		void load(FileGetter file) throws IOException {
			target = file.i();
			file.is(perGroup);
		}

	
	}

	
}
